src/jit/morph.cpp

```diff
@@ -17987,569 +17971,763 @@ GenTree* Compiler::fgMorphImplicitByRefArgs(GenTree* tree, bool isAddr)
-        // So let us be conservative and treat TYP_BYREF operations as AXC_IndWide and propagate a
-        // wider indirection context down the expr tree.
-        //
-        // Example, in unsafe code,
-        //
-        //   IL_000e  12 00             ldloca.s     0x0
-        //   IL_0010  7c 02 00 00 04    ldflda       0x4000002
-        //   IL_0015  12 00             ldloca.s     0x0
-        //   IL_0017  7c 01 00 00 04    ldflda       0x4000001
-        //   IL_001c  59                sub
-        //
-        // When visiting the GT_SUB node, if the types of either of the GT_SUB's operand are BYREF, then
-        // consider GT_SUB to be equivalent of an AXC_IndWide.
-        //
-        // Similarly for pointer comparisons and pointer escaping as integers through conversions, treat
-        // them as AXC_IndWide.
-        //
+            case GT_DYN_BLK:
+                assert(TopValue(2).Node() == node);
+                assert(TopValue(1).Node() == node->AsDynBlk()->Addr());
+                assert(TopValue(0).Node() == node->AsDynBlk()->gtDynamicSize);
 
-        case GT_CALL:
-        {
-            // Scan for byref args
-            GenTreeCall* const call = tree->AsCall();
-            for (GenTree* args = call->gtCallArgs; (args != nullptr); args = args->gtOp.gtOp2)
-            {
-                if (args->gtOp.gtOp1->gtType == TYP_BYREF)
+                // The block size may be the result of an indirection so we need
+                // to escape the location that may be associated with it.
+                EscapeValue(TopValue(0), node);
+
+                if (!TopValue(2).Indir(TopValue(1)))
                 {
-                    axcStack->Push(AXC_IndWide);
-                    return WALK_CONTINUE;
+                    // If the address comes from another indirection (e.g. DYN_BLK(IND(...))
+                    // then we need to escape the location.
+                    EscapeLocation(TopValue(1), node);
                 }
-            }
 
-            break;
-        }
+                PopValue();
+                PopValue();
+                break;
 
-        // BINOP
-        case GT_SUB:
-        case GT_MUL:
-        case GT_DIV:
-        case GT_UDIV:
-        case GT_OR:
-        case GT_XOR:
-        case GT_AND:
-        case GT_LSH:
-        case GT_RSH:
-        case GT_RSZ:
-        case GT_ROL:
-        case GT_ROR:
-        case GT_EQ:
-        case GT_NE:
-        case GT_LT:
-        case GT_LE:
-        case GT_GT:
-        case GT_GE:
-        case GT_ASG:
-        // UNOP
-        case GT_CAST:
-            if ((tree->gtOp.gtOp1->gtType == TYP_BYREF) ||
-                (tree->OperIsBinary() && (tree->gtOp.gtOp2->gtType == TYP_BYREF)))
-            {
-                axcStack->Push(AXC_IndWide);
-                return WALK_CONTINUE;
-            }
-            break;
+            default:
+                while (TopValue(0).Node() != node)
```

Comment:
We need to handle comma nodes explicitly. A comma node would "escape" the "value" produced by its first operand and "copy" the "value" produced by its second operand.

While I'm at it - casts are too missing. A cast from `TYP_BYREF` to `TYP_I_IMPL` would "copy" its input "address value". That would take care of some cases `(byte*)&sbyteLclVar` that currently end up making `sbyteLclVar` address exposed. But morph doesn't handle this well so it will have to wait.